### PR TITLE
Converting ChannelCache to be pluggable

### PR DESCRIPTION
**Pull request by panoramix360**
Work in progress PR resolving issue #355.

We made the adjustments to transform ChannelCache to be pluggable just as asked and created tests for this.

The PR is in Work in Progress because we need to update the Docs on the files created.

:)

### CODE REVIEW

**Comment by jchristgit** [comment]
Hi @panoramix360 and @gbottari, thanks for your contribution!

This looks great so far! I will give it a complete review once you consider it ready.

As for your question here:

> We noticed that there isn't any tests for ChannelCache so we are creating these but we had some problems with an ETS table that is needed for ChannelCache.get/1 that is only created on the Application.setup_ets_tables/0.

Due to the callbacks implementing the supervisor behaviour (just like in `Nostrum.Cache.GuildCache.ETS`), you should be able to skip the separate `Application.setup_ets_tables()` call, as the correct table should be created by `init`.

**Comment by gbottari** [comment]
Hi @jchristgit, thanks for your kind words!

I guess the problem we ran into is that `ChannelCache.ETS.lookup()` actually relies on `GuildCache.select_by()` behind the scenes. `select_by()` in turn, uses `Mapping.ChannelGuild.get_guild()` which relies on table `:channel_guild_map`. The only place table `:channel_guild_map` is initialized is in `Application.setup_ets_tables()`.

As we see it right now the only way to test ChannelCache is to initialize `:channel_guild_map` table somehow or maybe work around this dependency somehow.

**Comment by jchristgit** [comment]
Hi @gbottari. Of course, I completely missed that, sorry! Yeah, in that case, the current startup logic is perfectly fine. 👍

**Comment by panoramix360** [comment]
Hi @jchristgit, we finished the PR :)

Waiting for your approval.

:+1:

File: lib/nostrum/cache/channel_cache.ex
Comment (jchristgit) [inline lib/nostrum/cache/channel_cache.ex:22]:
```suggestion
  - The functions that the user calls, currently only `get/1` and `get!/1`
```

**Review by jchristgit — CHANGES_REQUESTED** [review]
Thanks for the PR! I've gotten around to testing this. We need to do two things:

- Remove the now obsolete creation of the `:channels` ETS table in `setup_ets_tables` in `lib/nostrum/application.ex`
- Add the channel cache to the `CacheSupervisor` children in `lib/nostrum/cache/cache_supervisor.ex`

**Comment by panoramix360** [comment]
@jchristgit thank you for your suggestions, I already changed and commited!

**Review by jchristgit — APPROVED** [review]
This all looks great now, thank you very much!

**Comment by jchristgit** [comment]
Thanks!

**Pull request merged**
